File: smart-toast/src/main/java/com/coder/zzq/smartshow/toast/SafeHandler.java
package com.coder.zzq.smartshow.toast;


import android.os.Handler;
import android.os.Message;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

final class SafeHandler extends Handler {
    private Handler mNestedHandler;
    private View mToastView;

    public SafeHandler(Handler nestedHandler, View toastView) {
        mNestedHandler = nestedHandler;
        mToastView = toastView;
    }

    @Override
    public void dispatchMessage(Message msg) {
        try {

            if (mToastView.getParent() != null && (mToastView.getParent() instanceof ViewGroup)) {
                ((ViewGroup) mToastView.getParent()).removeView(mToastView);
            }

            super.dispatchMessage(msg);

        } catch (WindowManager.BadTokenException | IllegalStateException ex) {
            //handle BadTokenException for android system bug on 7.1.x
            //handle IllegalStateException for some devices
        }
    }

    @Override
    public void handleMessage(Message msg) {
        mNestedHandler.handleMessage(msg);
    }
}

File: smart-toast/src/main/java/com/coder/vincent/smart_toast/compact/SafeHandler.kt
package com.coder.vincent.smart_toast.compact

import android.os.Handler
import android.os.Looper
import android.os.Message

internal class SafeHandler(private val mNestedHandler: Handler) :
    Handler(Looper.getMainLooper()) {
    override fun dispatchMessage(msg: Message) {
        kotlin.runCatching {
            super.dispatchMessage(msg)
        }
    }

    override fun handleMessage(msg: Message) {
        mNestedHandler.handleMessage(msg)
    }
}